src/beamlime/applications/_nexus_helpers.py

```diff
@@ -1,155 +1,365 @@
 # SPDX-License-Identifier: BSD-3-Clause
 # Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
-from typing import Iterable, Mapping, Optional, Tuple
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    Literal,
+    Mapping,
+    Optional,
+    Sequence,
+    TypedDict,
+)
 
 import numpy as np
 
 
-def _dtype(value):
-    if hasattr(value, 'dtype'):
-        return str(value.dtype)
-    if isinstance(value, str):
-        return 'string'
-    if isinstance(value, int):
-        return 'int'
-    if isinstance(value, float):
-        return 'float'
-    return 'unknown'
-
-
-def _init_dataset(group, field, value):
-    dataset = find_nexus_structure(group, (field,))
-    if dataset is None:
-        dataset = dict(
-            module='dataset',
-            config=dict(
-                name=field,
-                dtype=_dtype(value),
-                values=value,
-            ),
+class _NexusDataset(TypedDict):
+    """``dataset`` module structure in the nexus json format."""
+
+    module: str
+    config: dict[str, Any]
+    attributes: list[Mapping[str, Any]]
+
+
+NexusPath = tuple[str | None, ...]
+
+
+class DataModuleParent(TypedDict):
+    """``group`` that holds the data module place holder or datasets
+
+    It was named as a parent, not a module
+    since the module place holder is one of the children of it.
+    """
+
+    children: list[Mapping]
+    """A module place holder or datasets."""
+    name: str
+    """Name of the group."""
+
+
+DataModuleStore = dict[NexusPath, DataModuleParent]
```

Comment:
I think this should be called `ModuleStore` or `NexusGroupStore`.
`DataModule` is more convoluted than simply `Module`, but it doesn't really communicate more than just `Module`. `NexusGroupStore` is longer, but explains a bit more what the type is.

If we want to let the name indicate that it's a store for the streams then  maybe `StreamStore` or `StreamedGroupStore` are other good options.